### PR TITLE
JAMES-3006 Add a TaskFactory for modular task registration in webadmin

### PR DESCRIPTION
**Pull request by chibenwa**
Follow up of https://github.com/linagora/james-project/pull/2988

### CODE REVIEW

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/tasks/TaskFactory.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+import com.github.steveash.guavate.Guavate;
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import spark.Request;
+import spark.Route;
+
+public class TaskFactory implements TaskGenerator {
```
Comment (remk) [inline server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/tasks/TaskFactory.java]: I think that the name is misleading.
s/TaskFactory/TaskFromRequestRegistry ?

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/tasks/TaskGenerator.java [hunk 2]
```diff
@@ -17,22 +17,37 @@
...
-class ReIndexingRoutesUtil {
-    static void enforceTaskParameter(Request request) {
-        String task = request.queryParams("task");
-        if (!"reIndex".equals(task)) {
-            throw ErrorResponder.builder()
-                .statusCode(HttpStatus.BAD_REQUEST_400)
-                .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
-                .message("task query parameter is mandatory. The only supported value is `reIndex`")
-                .haltError();
+public interface TaskGenerator {
```
Comment (remk) [inline server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/tasks/TaskGenerator.java]: s/TaskGenerator/TaskFromRequest ?

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/tasks/TaskGenerator.java [hunk 3]
```diff
@@ -17,22 +17,37 @@
...
+
+        @Override
+        public Object handle(Request request, Response response) throws Exception {
+            Task task = taskGenerator.generate(request);
+            TaskId taskId = taskManager.submit(task);
+            return TaskIdDto.respond(response, taskId);
         }
     }
+
+    Task generate(Request request) throws Exception;
```
Comment (remk) [inline server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/tasks/TaskGenerator.java]: s/generate/from ?

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ReindexingRoutesTest.java [hunk 4]
```diff
@@ -141,7 +142,8 @@ void fullReprocessingShouldFailWithBadTask() {
                     .statusCode(HttpStatus.BAD_REQUEST_400)
                     .body("statusCode", is(400))
                     .body("type", is(ErrorResponder.ErrorType.INVALID_ARGUMENT.getType()))
-                    .body("message", is("task query parameter is mandatory. The only supported value is `reIndex`"));
+                    .body("message", is("Invalid arguments supplied in the user request"))
+                    .body("details", is("Invalid value supplied for 'task': bad. Supported values are [reIndex]"));
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ReindexingRoutesTest.java]: that's the parameter name supplied...

**Comment by chibenwa** [comment]
(I will wait tomorrow for naming related issues, to see if some alternative proposal pops in!)

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java [hunk 5]
```diff
@@ -493,32 +495,21 @@ private Task toAllMailReprocessingTask(Request request) throws UnsupportedEncodi
         @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Bad request - unknown action")
     })
     public void defineReprocessOne() {
-        service.patch(MAIL_REPOSITORIES + "/:encodedPath/mails/:key", (request, response) -> {
-            Task task = toOneMailReprocessingTask(request);
-            TaskId taskId = taskManager.submit(task);
-            return TaskIdDto.respond(response, taskId);
-        }, jsonTransformer);
+        service.patch(MAIL_REPOSITORIES + "/:encodedPath/mails/:key",
+            TaskFactory.builder()
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: could we have a shortcut for single-entry task factories?

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 6]
```diff
@@ -1075,7 +1075,8 @@ public void reprocessingAllTaskShouldRejectInvalidActions() {
             .statusCode(HttpStatus.BAD_REQUEST_400)
             .body("statusCode", is(400))
             .body("type", is(ErrorResponder.ErrorType.INVALID_ARGUMENT.getType()))
-            .body("message", is("action query parameter is mandatory. The only supported value is `reprocess`"));
+            .body("message", is("Invalid arguments supplied in the user request"))
+            .body("details", is("Invalid value supplied for 'action': invalid. Supported values are [reprocess]"));
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: you loose the "query parameter" wording that was important.

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/tasks/TaskGenerator.java [hunk 7]
```diff
@@ -38,10 +42,12 @@
         }
 
         @Override
-        public Object handle(Request request, Response response) throws Exception {
+        public TaskIdDto handle(Request request, Response response) throws Exception {
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/tasks/TaskGenerator.java]: :+1:

**Comment by Arsnael** [comment]
Rebased to resolve conflicts with master branch

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**